### PR TITLE
add descriptor loader

### PR DESCRIPTION
**Pull request by knoguchi**
This is a preliminary work for the  #78 "create ProtoSchema from descriptor file".

A "descriptor file" is a set of descriptors.  A descriptor is a binary representation of the .proto file.  Typically one (or more) of the descriptors is the main, and the others are dependencies.

This reader creates a `NativeProtobufSchema` object from the main descriptor.  Dependencies are "copied" into the main descriptor before generating the object.  I chose copying over implementing the import mechanism per #86 discussion. The import statement is a potential security risk, and it would certainly add the complexity.  

This descriptor reader should work for simple message types that do not have references to other nested message types.  #73 needs to be fixed.

Note: descriptor.proto was [copied from Google Protobuf](https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto).   The `reserved` fields were commented out because ProtoParser could not parse.  See #84.

### CODE REVIEW

File: protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReader.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        InputStream in = getClass().getResourceAsStream(DESCRIPTOR_PROTO);
+        descriptorFileSchema = ProtobufSchemaLoader.std.load(in, "FileDescriptorSet");
+        in.close();
+    }
+
+    public Map<String, ProtoFile> readFileDescriptorSet(String descriptorFilePath) throws IOException
+    {
+        InputStream fin;
+        fin = this.getClass().getClassLoader().getResourceAsStream(descriptorFilePath);
+        if (fin == null) {
```
Comment (cowtowncoder) [inline protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReader.java]: If the resource can not be read, why would accessing it as `URL` work differently (and should it)? Maybe at least add a comment explaining what is the motivation here

File: protobuf/src/test/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+{
+    public void testParsing() throws Exception
+    {
+        DescriptorReader reader = new DescriptorReader();
+        assertNotNull(reader);
+        Map<String, ProtoFile> protoFileMap = reader.readFileDescriptorSet("main.desc");
+        NativeProtobufSchema nativeSchema = NativeProtobufSchema.construct(protoFileMap.get("main.proto"));
+
+        ProtobufSchema schema = nativeSchema.forType("main1");
+        System.out.println(schema.toString());
```
Comment (cowtowncoder) [inline protobuf/src/test/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReaderTest.java]: This should be removed.

**Comment by cowtowncoder** [comment]
Ok, I think this could be a good starting point. There are some things I'd probably change, such as making `ProtobufMapper` initiate work instead of reading things in constructor (which is typically not something to avoid), but in general I think it makes sense. This would also remove the dependency from reader to constructing mapper which seems wrong way around.
I think it also would make sense to take different kinds of input sources: assuming a String defines a resource path seems limiting; it may be one option but I assume there are other cases such as reading from an external file.

But one question first: what is intended usage pattern for `ProtoFile`s? I'd prefer not exposing 3rd party types as-is through API, if they can be encapsulated somehow. Can they be converted to protoc definition? Test seems to just verify that something was read, but I assume these are useful for some actual usage.

File: protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReader.java (hunk 1 above)
Comment (knoguchi) [inline protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorReader.java]: I was trying to make it dual purpose -- load the descriptor from resources directory as well as external URL.    I will remove the resources.  I guess the descriptors in the resources directory is only useful for unit tests.

**Comment by knoguchi** [comment]
Thanks for the comments.   The reason why I exposed `ProtoFile`s map is to let the driver to pick the main proto file, then call `NativeProtobufSchema.construct(ProtoFile)` API.  
I will modify it to be something like below to hide the 3rd party API.

```
// in the DescriptorReader class
private Map<String, ProtoFile> protoFileMap;

NativeProtobufSchema forProto(String protoName) {
    return protoFileMap.get(protoName);
}
```

`ProtoFile.toSchema()` method can generate the .proto string but I did not intend the usage.

The test case barely shows that the two .proto files (main.proto and other.proto) are merged.  I will add tests that actually check the correctness of the output.

Other items that you pointed out:
1. I will move the ProtobufMapper for the descriptor schema out of the reader class.
2. add input sources:
    URL
    InputStream
    byte[]

**Comment by cowtowncoder** [comment]
Unfortunately haven't had time to spend on this, but wanted to add a note to say it's still on my radar.

File: protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorLoader.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            }
+            return descriptorMapper.readerFor(FileDescriptorSet.class)
+                                   .with(descriptorFileSchema)
+                                   .readValue(in);
+        }
+        finally {
+            try {
+                in.close();
+            }
+            catch (IOException e) {
```
Comment (cowtowncoder) [inline protobuf/src/main/java/com/fasterxml/jackson/dataformat/protobuf/schema/DescriptorLoader.java]: Why catch? Shouldn't it just be propagated as-is?

**Comment by cowtowncoder** [comment]
Actually I think I can make some changes here from API perspective, but retain functionality.
Nothing big, just bit of touch up.

**Comment by cowtowncoder** [comment]
Ok so did refactoring, which is sizable by lines of code. But didn't much change functionality: mostly just changes access point, which is now via `ProtobufMapper`.

**Pull request merged**